runtime/lua/vim/iter.lua

```diff
@@ -179,6 +194,59 @@ function ListIter.filter(self, f)
   return self
 end
 
+--- Flattens an |list-iterator| pipeline, un-nesting nested values.
+--- The flattened list will exclude nil, and any values with non-numeric keys
+---
+--- Example:
+---
+--- ```lua
+--- vim.iter({1, {2}, {{3}}}):flat():totable()
+--- ```
+--- (returns `{1, 2, {3}}`)
+---
+--- ```lua
+--- vim.iter({1, {a = 2}, {nil, {nil}}, {{3}}}):flat():totable()
+--- ```
+--- (returns `{1, {}, 3}`)
+---
+--- ```lua
+--- vim.iter({[2] = 2}):flat():totable()
+--- vim.iter({nil, 2})
+---
+--- ```
+--- Both give an error because the iterator being flattened is not list-like
+--- (the indices in the source table are not consecutive)
```

Comment:
I don't think we need to explain what "list-like" means here. There are already other places that do that, and if needs to be repeated there are more appropriate locations. For the this function's docstring, assume that "only works for list-like tables" is well understood. No need to provide anti-examples.